python/pyarrow/scalar.pxi

```diff
@@ -1173,7 +1173,8 @@ cdef class MapScalar(ListScalar, Mapping):
         if not maps_as_pydicts:
             return list(self)
         result_dict = {}
-        for key, value in self:
+        for k, v in zip(self.values.field(self.type.key_field.name), self.values.field(self.type.item_field.name)):
+            key = k.as_py()
```

Comment:
I think we can simplify this a bit:

```suggestion
        for key, value in zip(self.keys(), self.values.field(self.type.item_field.name)):
```

and maybe something for the second iterator?